x/wasm/IBC.md

```diff
@@ -0,0 +1,281 @@
+* When receiving an Ack or Error packet, the contract also receives the
+  original packet that it set on Send 
+  
+  
+## Workflow
+
+Establishing *Clients* and *Connections* is out of the scope of this
+module and must be created by the same means as for `ibc-transfer`
+(via the cli or otherwise). `x/wasm` will bind a unique *Port* for each
+"IBC Enabled" contract.
+
+An "IBC Enabled" contract may dispatch the following messages not available
+to other contracts:
+
+* `IBCSendMsg` - this sends an IBC packet over an established channel. 
+* `IBCOpenChannel` - given a ConnectionID and a remote Port, this will request
+  to open a new channel with the given chain
+* `IBCCloseChannel` - given an existing channelID bound to this contract's Port,
+  initiate the closing sequence and reject all pending packets.
+
+They are returned from `handle` just like any other `CosmosMsg`
+(For mocks, we will trigger this externally, later only valid contract addresses 
+should be able to do so).
+
+An "IBC Enabled" contract must support the following callbacks from the runtime
+(we will likely multiplex many over one wasm export, as we do with handle, but these 
+are the different calls we must support):
+
+Packet Lifecycle:
+
+* `IBCRecvPacket` - when another chain sends a packet destined to
+  an IBCEnabled contract, this will be routed to the proper contract
+  and call an exposed function for this purpose.
+* `IBCPacketAck` - The original sender of `IBCSendMsg` will
+  get this callback eventually if the message was
+  processed on the other chain (this may be either a success or an error,
+  but comes from the app-level protocol, not the IBC protocol).
+* `IBCPacketDropped` - The original sender of `IBCSendMsg` will 
+  get this callback eventually if the message failed to be
+  processed on the other chain (for timeout, closed channel, or 
+  other IBC-level failure)
+  
+Note: We may add some helpers inside the contract to map `IBCPacketAck` / `IBCPacketDropped`
+to `IBCPacketSucceeded` / `IBCPacketFailed` assuming they use the standard envelope. However,
+we decided not to enforce this on the Go-level, to allow contracts to communicate using protocols
+that do not use this envelope.
+
+Channel Lifecycle Hooks:
+
+* `OnChanOpenTry` - this is called when another chain attempts to open
+  a channel with this contract. It provides the protocol versions that
+  the initiate expects, and the contract can reject it or accept it
+  and return the protocol version it will communicate with.
+* `OnChanOpenConfirm` - this is called when the other party has agreed
+  to the channel and all negotiation is over. This will be called exactly
+  once for any established channel, and the contract should record these
+  open channels in the internal state.
+* `OnChanClosed` - this is called when an existing channel is closed
+  for any reason, allowing the client to update the state there.
+  This will (likely? @cwgoes?) be accompanied by `IBCPacketFailed`
```

Comment:
Do you mean `IBCPacketDropped`? Yes, packets can be timed-out after a channel is closed